ext/zlib/zlib.c

```diff
@@ -2921,41 +2946,42 @@ gzfile_readpartial(struct gzfile *gz, long len, VALUE outbuf)
             return outbuf;
         }
     }
-    while (!ZSTREAM_IS_FINISHED(&gz->z) && ZSTREAM_BUF_FILLED(&gz->z) == 0) {
+
+    // fill stream until the buffer can be filled. while this isn't strictly
+    // necessary, it does help cut down on the number of calls to readpartial
+    // in the typical case.
+    while (!ZSTREAM_IS_FINISHED(&gz->z) && ZSTREAM_BUF_FILLED(&gz->z) < len) {
```

Comment:
This is probably the bug I was seeing. It wasn't reading enough to fill the len requested, so it was returning less than requested even when the amount was available.

Reply:
Isn't that what `readpartial` means?

Reply:
My understanding is that readpartial reads as much as it can, up to the limit, unless it would block.

Reply:
yes, i fixed now

Reply:
I was also mistaken about the exact behavior of readpartial. Sorry for any confusion. I double checked this branch myself locally and this definitely solves #56.